pkg/router/apisix.go

```diff
@@ -0,0 +1,221 @@
+		Weight:             &canaryWeight,
+		Subset:             primaryBackend.Subset,
+	}
+
+	targetHttpRoute.Backends = append(targetHttpRoute.Backends, canaryBackend)
+	apisixRouteClone.Spec.HTTP[targetIndex] = *targetHttpRoute
+
+	canaryApisixRouteName := fmt.Sprintf("%s-%s-canary", canary.Spec.RouteRef.Name, apexName)
+	canaryApisixRoute, err := ar.apisixClient.ApisixV2().ApisixRoutes(canary.Namespace).Get(context.TODO(), canaryApisixRouteName, metav1.GetOptions{})
+
+	if errors.IsNotFound(err) {
+		route := &a6v2.ApisixRoute{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:        canaryApisixRouteName,
+				Namespace:   canary.Namespace,
+				Annotations: apisixRouteClone.Annotations,
+				Labels:      apisixRouteClone.Labels,
+			},
+			Spec: apisixRouteClone.Spec,
+		}
+
+		if ar.setOwnerRefs {
+			route.OwnerReferences = []metav1.OwnerReference{
+				*metav1.NewControllerRef(canary, schema.GroupVersionKind{
+					Group:   flaggerv1.SchemeGroupVersion.Group,
+					Version: flaggerv1.SchemeGroupVersion.Version,
+					Kind:    flaggerv1.CanaryKind,
+				}),
+			}
+		}
+
+		_, err := ar.apisixClient.ApisixV2().ApisixRoutes(canary.Namespace).Create(context.TODO(), route, metav1.CreateOptions{})
+		if err != nil {
+			return fmt.Errorf("APISIX route %s.%s create error: %w", route.Name, route.Namespace, err)
+		}
+
+		ar.logger.With("canary", fmt.Sprintf("%s.%s", canary.Name, canary.Namespace)).
+			Infof("APISIX route %s.%s created", route.GetName(), canary.Namespace)
+		return nil
+	} else if err != nil {
+		return fmt.Errorf("APISIX route %s.%s query error: %w", canaryApisixRouteName, canary.Namespace, err)
+	}
+
+	if diff := cmp.Diff(canaryApisixRoute.Spec, apisixRouteClone.Spec,
+		cmpopts.IgnoreFields(a6v2.ApisixRouteHTTPBackend{}, "Weight")); diff != "" {
+		iClone := canaryApisixRoute.DeepCopy()
+		iClone.Spec = apisixRouteClone.Spec
+
+		_, err := ar.apisixClient.ApisixV2().ApisixRoutes(canary.Namespace).Update(context.TODO(), iClone, metav1.UpdateOptions{})
+		if err != nil {
+			return fmt.Errorf("APISIX route %s.%s update error: %w", canaryApisixRouteName, iClone.Namespace, err)
+		}
+		ar.logger.With("canary", fmt.Sprintf("%s.%s", canary.Name, canary.Namespace)).
+			Infof("Apisix route %s updated", canaryApisixRouteName)
+	}
+
+	return nil
+}
+
+// GetRoutes returns the destinations weight for primary and canary
+func (ar *ApisixRouter) GetRoutes(canary *flaggerv1.Canary) (
+	primaryWeight int,
+	canaryWeight int,
+	mirrored bool,
+	err error,
+) {
+	apexName, primaryName, _ := canary.GetServiceNames()
+	canaryApisixRouteName := fmt.Sprintf("%s-%s-canary", canary.Spec.RouteRef.Name, apexName)
+	apisixRoute, err := ar.apisixClient.ApisixV2().ApisixRoutes(canary.Namespace).Get(context.TODO(), canaryApisixRouteName, metav1.GetOptions{})
+	if err != nil {
+		err = fmt.Errorf("apisix route %s.%s query error: %w", canaryApisixRouteName, canary.Namespace, err)
+		return
+	}
+
+	for _, backend := range apisixRoute.Spec.HTTP[0].Backends {
```

Comment:
i don't think we can assume the object to be at the 0th index, because of
```go
apisixRouteClone.Spec.HTTP[targetIndex] = *targetHttpRoute
```
in `Reconcile()` above

Reply:
done